### PR TITLE
Automate TPL callback registrations

### PR DESCRIPTION
**Pull request by jsiirola**
## Summary/Motivation:
This PR updates how we handle registration of `numpy` types as native numeric types.  This updates the `attempt_import` machinery to hook into Python's regular `importlib` process.  Now deferred import modules with callbacks (including `numpy`) are registered with a custom module finder that ensures that if they are imported (even by code outside of Pyomo), the post-import callback is triggered.

## Changes proposed in this PR:
- add `DeferredImportCallbackFinder` / `DeferredImportCallbackLoader` to hook into `importlib` and endure deferred import modules with callbacks always have their callbacks called
- rename `attempt_import(defer_check=` to `attempt_import(defer_import=`
- by default `attempt_import` will immediately process modules that have already been imported 

### Legal Acknowledgement

By contributing to this software project, I have read the [contribution guide](https://pyomo.readthedocs.io/en/stable/contribution_guide.html) and agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the BSD license.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

### CODE REVIEW

File: pyomo/common/dependencies.py [hunk 1]
```diff
@@ -433,6 +442,83 @@ def check_min_version(module, min_version):
...
+        )
+        return spec
+
+    def invalidate_caches(self):
+        pass
+
+
+_DeferredImportCallbackFinder = DeferredImportCallbackFinder()
+# Insert the DeferredImportCallbackFinder at the beginning of the
+# sys.meta_path to that it is found before the standard finders (so that
```
Comment (mrmundt) [inline pyomo/common/dependencies.py:516]:
```suggestion
# sys.meta_path so that it is found before the standard finders (so that
```

**Review by mrmundt — APPROVED** [review]
Smallest of NFC changes, but it's good besides that.

File: pyomo/common/dependencies.py [hunk 2]
```diff
@@ -842,41 +1034,48 @@ def _pyutilib_importer():
...
-# not installed as a Framework (as is the case in the CI systems)
-matplotlib, matplotlib_available = attempt_import(
-    'matplotlib',
-    callback=_finalize_matplotlib,
-    deferred_submodules=['pyplot', 'pylab'],
-    catch_exceptions=(ImportError, RuntimeError),
-)
+#
+# Note: because we will be calling
+# declare_deferred_modules_as_importable, it is important that the
```
Comment (blnicho) [inline pyomo/common/dependencies.py:1039]:
```suggestion
# declare_modules_as_importable, it is important that the
```

**Review by blnicho — APPROVED** [review]
I found one typo but otherwise this looks great

File: pyomo/common/dependencies.py (hunk 2 above)
Comment (jsiirola) [reply pyomo/common/dependencies.py]: Actually, this whole comment is not longer relevant after switching over to using the context manager.  Removed.

**Pull request merged**